### DISCUSSION TITLE
Custom training with some onnx models

### DISCUSSION THREAD

**Discussion by IzanCatalan** [microsoft/onnxruntime#16598] [Other Q&A]
Hi everyone, I would like to know if it is possible to define a custom loss function together with a custom forward and backward propagation methods to modify the weights to force some of them to be a determinate value?

Thanks.

**Comment by baijumeswani** [comment]
ONNX Runtime supports two forms of training:

1. [On-Device Training](https://onnxruntime.ai/docs/api/python/on_device_training/overview.html)
2. [Large Model Training](https://onnxruntime.ai/docs/api/python/ortmodule/overview.html)

With large model training with ORTModule, the input must be a PyTorch model. You can define the PyTorch model (forward and loss) however you deem fit for your application.

With on-device training, the ONNX Runtime training API takes in an input ONNX training model. This training model must be generated using the ONNX Runtime offline artifact generation tools. With these tools, you could pass in a forward ONNX model, with a custom loss logic, and the tool will automatically build the gradient graph.
